fix(api): return NULL when NULL is passed to `Array.zip`

NickCrews opened the pull request:
fixes https://github.com/ibis-project/ibis/issues/8650

NickCrews commented on ibis/backends/duckdb/compiler.py:
```diff
@@ -187,14 +187,17 @@ def visit_ArrayZip(self, op, *, arg):
             func,
         )
+        # if any of the input arrays in arg are NULL, the result is NULL
+        any_arg_null = reduce(lambda x, y: x | y.is_(NULL), arg, False)
```
this `reduce` was an uniformed acceptance from a github copilot suggestion (that worked, much wow), so please verify this is a good way to accomplish this.

cpcloud replied:
> please verify this is a good way to accomplish this

lol at AI

```suggestion
        any_arg_null = reduce(lambda x, y: x | y.is_(NULL), arg, FALSE)
```

The `FALSE` may need to be imported

NickCrews replied:
`sg.or_()` actually is what we want I think

NickCrews commented on ibis/backends/duckdb/compiler.py:
```diff
@@ -187,14 +187,17 @@ def visit_ArrayZip(self, op, *, arg):
         func = sge.Lambda(this=body, expressions=[i])
-        return self.f.list_apply(
+        regular = self.f.list_apply(
             self.f.range(
```
could instead put the null check up here, so that the input to `list_apply` is NULL. IDK if that is any more performant...

cpcloud replied:
I think it's fine as is.

NickCrews commented on ibis/backends/tests/test_array.py:
```diff
@@ -850,6 +850,15 @@ def test_zip(backend):
     assert len(x[0]) == len(s[0])
 
+    # clickhouse nested types can't be null
+    if backend.name() != "clickhouse":
```
was lazy about using pytest connection fixtures, etc. I can split this up better if you want, but this isn't that ugly to me.

cpcloud commented on ibis/expr/types/arrays.py:
```diff
@@ -930,43 +930,41 @@ def zip(self, other: ArrayValue, *others: ArrayValue) -> ArrayValue:
         Array
             Array of structs where each struct field is an element of each input
             array.
+            The fields are named `f1`, `f2`, `f3`, etc.
```
nit: can you put this on the same line as the previous sentence?

NickCrews commented:
@cpcloud thanks for the review! all good suggestions, addressed them all. Check that you like the way I split out the new test, reusing a single decorator to avoid repeating myself.

cpcloud commented on ibis/backends/tests/test_array.py:
```diff
@@ -818,8 +818,7 @@ def test_unnest_struct(con):
 
-@builtin_array
-@pytest.mark.notimpl(
+ARRAY_ZIP_NOTIMPL = pytest.mark.notimpl(
```
```suggestion
array_zip_notimpl = pytest.mark.notimpl(
```

cpcloud commented on ibis/backends/tests/test_array.py:
```diff
@@ -833,6 +832,10 @@ def test_unnest_struct(con):
+
+
+@builtin_array
+@ARRAY_ZIP_NOTIMPL
```
```suggestion
@array_zip_notimpl
```

cpcloud commented on ibis/backends/tests/test_array.py:
```diff
@@ -851,6 +854,22 @@ def test_zip(backend):
 
 
+@builtin_array
+@ARRAY_ZIP_NOTIMPL
```
```suggestion
@array_zip_notimpl
```

cpcloud reviewed (approved):
LGTM, just some suggestions that I will add via the GitHub UI.

cpcloud commented:
Interesting, looks like the backends that support zip (except clickhouse) all have this behavior by default.

cpcloud commented:
That said, the DuckDB implementation isn't calling a specific duckdb `zip` function, so it's kind of up to us to make it behave consistently.

cpcloud reviewed (approved):
Thanks!
